File: spring-web/src/main/java/com/github/solayw/webutil/jpa/EntityManagerDelegate.java
package com.github.solayw.webutil.jpa;


import com.github.solayw.webutil.Page;

import java.util.List;
import java.util.Objects;

import javax.persistence.EntityManager;
import javax.persistence.Query;

import lombok.experimental.Delegate;


public class EntityManagerDelegate implements EntityManager
{
    @Delegate
    private EntityManager entityManager;

    public EntityManagerDelegate(EntityManager entityManager) {
        this.entityManager = entityManager;
    }


    public static Object singleOrNull(Query query) {
        final List list = query.getResultList();
        if(list.size() > 0) {
            return list.get(0);
        }
        return null;
    }

    public <T> PageList<T> pageQuery(PageQueryObj<T> pageQueryObj) {
        final List<T> list = pageQueryObj.queryForList(this);
        final int count = pageQueryObj.queryForCount(this);
        PageList<T> pageList = new PageList<>(list);
        pageList.total = count;
        Page page = Objects.requireNonNull(pageQueryObj.getPage());
        int totalPage = count / page.getPageSize();
        if(totalPage * page.getPageSize() != count) {
            totalPage += 1;
        }
        pageList.totalPage = totalPage;
        return pageList;
    }


}

File: spring-web/src/main/java/com/github/solayw/webutil/jpa/EntityManagerDelegate.kt
package com.github.solayw.webutil.jpa

import java.util.*
import javax.persistence.EntityManager
import javax.persistence.Query


class EntityManagerDelegate(private val entityManager: EntityManager) : EntityManager by entityManager {
    fun <T> pageQuery(pageQueryObj: PageQueryObj<T>): PageList<T> {
        val list = pageQueryObj.queryForList(this)
        val count = pageQueryObj.queryForCount(this)
        val pageList = PageList(list)
        pageList.total = count
        val page = Objects.requireNonNull(pageQueryObj.page)
        var totalPage = count / page.pageSize
        if (totalPage * page.pageSize != count) {
            totalPage += 1
        }
        pageList.totalPage = totalPage
        return pageList
    }

    companion object {
        fun singleOrNull(query: Query): Any? {
            val list = query.resultList
            return if (list.size > 0) {
                list[0]
            } else null
        }
    }
}
